// @embrace-io/web-sdk/build/cjs/resources/webSdkResource.cjs
'use strict';

var resources = require('@opentelemetry/resources');
var semanticConventions = require('@opentelemetry/semantic-conventions');
var index = require('./constants/index.cjs');
var appInstanceId = require('./appInstanceId.cjs');
var attributes = require('../constants/attributes.cjs');

const getWebSDKResource = ({ diagLogger, appVersion, pageSessionStorage })=>{
    /* We need to trim the app  version to remove any leading/trailing spaces
    added by our cli tool. This is required to guarantee that the version is always
    20 characters long in the final bundle, so sourcemaps don't get confused by
    changing the length of the inlined app version. E.g. versions "0.0.1" and "0.0.115"
    are both valid, but injecting them without leading whitespaces will result in
    different lengths, pushing the sourcemaps mapping out of range. Instead,
    "               0.0.1" and "             0.0.115" are both 20 characters long,
    and we trim them before loading at runtime

    if the app version has been passed as an argument, we use it instead of the
    one processed by our cli tool. This will allow for no cli installations,
    like importing the web sdk from a CDN
    */ const processedAppVersion = appVersion ?? index.TEMPLATE_APP_VERSION.trim();
    return resources.resourceFromAttributes({
        [semanticConventions.ATTR_SERVICE_NAME]: index.EMBRACE_SERVICE_NAME,
        [semanticConventions.ATTR_TELEMETRY_SDK_NAME]: index.EMBRACE_SERVICE_NAME,
        app_version: processedAppVersion,
        app_framework: index.NATIVE_FRAMEWORK,
        sdk_version: index.SDK_VERSION,
        [semanticConventions.ATTR_TELEMETRY_SDK_VERSION]: index.SDK_VERSION,
        sdk_simple_version: 1,
        sdk_platform: 'web',
        [semanticConventions.ATTR_TELEMETRY_SDK_LANGUAGE]: 'webjs',
        [attributes.KEY_EMB_APP_INSTANCE_ID]: appInstanceId.getAppInstanceId(pageSessionStorage, diagLogger),
        [semanticConventions.ATTR_USER_AGENT_ORIGINAL]: window.navigator.userAgent,
        screen_resolution: `${window.screen.width}x${window.screen.height}`
    });
};

exports.getWebSDKResource = getWebSDKResource;
