// @embrace-io/web-sdk/build/cjs/constants/attributes.cjs
'use strict';

const KEY_EMB_TYPE = 'emb.type';
const KEY_EMB_STATE = 'emb.state';
const KEY_EMB_COLD_START = 'emb.cold_start';
const KEY_EMB_SESSION_NUMBER = 'emb.session_number';
const KEY_EMB_SDK_STARTUP_DURATION = 'emb.sdk_startup_duration';
const KEY_PREFIX_EMB_PROPERTIES = 'emb.properties.';
const KEY_EMB_SESSION_REASON_ENDED = 'emb.session_end_type';
const KEY_EMB_SESSION_REASON_STARTED = 'emb.session_start_type';
const KEY_EMB_JS_EXCEPTION_STACKTRACE = 'emb.stacktrace.js';
const KEY_EMB_EXCEPTION_HANDLING = 'emb.exception_handling';
const KEY_EMB_ERROR_CODE = 'emb.error_code';
const KEY_EMB_APP_INSTANCE_ID = 'emb.app_instance_id';
const KEY_EMB_TAB_ID = 'emb.tab_id';
const KEY_EMB_PARENT_TAB_ID = 'emb.parent_tab_id';
const KEY_EMB_EXPERIENCE_ID = 'emb.experience_id';
const KEY_EMB_ERROR_LOG_COUNT = 'emb.error_log_count';
const KEY_EMB_INSTRUMENTATION = 'emb.instrumentation';
const KEY_EMB_UNHANDLED_EXCEPTIONS_COUNT = 'emb.unhandled_exceptions_count';
const KEY_EMB_FROM_STORAGE = 'emb.from_storage';
const KEY_EMB_JS_FILE_BUNDLE_IDS = 'emb.js_file_bundle_ids';
const KEY_EMB_W3C_TRACEPARENT = 'emb.w3c_traceparent';
const KEY_VIEW_NAME = 'view.name';
var EMB_TYPES = /*#__PURE__*/ function(EMB_TYPES) {
    EMB_TYPES["Session"] = "ux.session";
    EMB_TYPES["View"] = "ux.view";
    EMB_TYPES["Network"] = "perf.network_request";
    EMB_TYPES["Perf"] = "perf";
    EMB_TYPES["SystemLog"] = "sys.log";
    EMB_TYPES["SystemException"] = "sys.exception";
    EMB_TYPES["WebVital"] = "ux.web_vital";
    EMB_TYPES["ResourceFetch"] = "ux.resource_fetch";
    EMB_TYPES["DocumentLoad"] = "ux.document_load";
    return EMB_TYPES;
}({});
var EMB_STATES = /*#__PURE__*/ function(EMB_STATES) {
    EMB_STATES["Foreground"] = "foreground";
    EMB_STATES["Background"] = "background";
    return EMB_STATES;
}({});
var EMB_NAVIGATION_INSTRUMENTATIONS = /*#__PURE__*/ function(EMB_NAVIGATION_INSTRUMENTATIONS) {
    EMB_NAVIGATION_INSTRUMENTATIONS["DeclarativeLegacy"] = "react_router_declarative_legacy";
    EMB_NAVIGATION_INSTRUMENTATIONS["Declarative"] = "react_router_declarative";
    EMB_NAVIGATION_INSTRUMENTATIONS["Data"] = "react_router_data";
    return EMB_NAVIGATION_INSTRUMENTATIONS;
}({});
var EMB_ERROR_INSTRUMENTATIONS = /*#__PURE__*/ function(EMB_ERROR_INSTRUMENTATIONS) {
    EMB_ERROR_INSTRUMENTATIONS["ReactErrorBoundary"] = "react_error_boundary";
    return EMB_ERROR_INSTRUMENTATIONS;
}({});

exports.EMB_ERROR_INSTRUMENTATIONS = EMB_ERROR_INSTRUMENTATIONS;
exports.EMB_NAVIGATION_INSTRUMENTATIONS = EMB_NAVIGATION_INSTRUMENTATIONS;
exports.EMB_STATES = EMB_STATES;
exports.EMB_TYPES = EMB_TYPES;
exports.KEY_EMB_APP_INSTANCE_ID = KEY_EMB_APP_INSTANCE_ID;
exports.KEY_EMB_COLD_START = KEY_EMB_COLD_START;
exports.KEY_EMB_ERROR_CODE = KEY_EMB_ERROR_CODE;
exports.KEY_EMB_ERROR_LOG_COUNT = KEY_EMB_ERROR_LOG_COUNT;
exports.KEY_EMB_EXCEPTION_HANDLING = KEY_EMB_EXCEPTION_HANDLING;
exports.KEY_EMB_EXPERIENCE_ID = KEY_EMB_EXPERIENCE_ID;
exports.KEY_EMB_FROM_STORAGE = KEY_EMB_FROM_STORAGE;
exports.KEY_EMB_INSTRUMENTATION = KEY_EMB_INSTRUMENTATION;
exports.KEY_EMB_JS_EXCEPTION_STACKTRACE = KEY_EMB_JS_EXCEPTION_STACKTRACE;
exports.KEY_EMB_JS_FILE_BUNDLE_IDS = KEY_EMB_JS_FILE_BUNDLE_IDS;
exports.KEY_EMB_PARENT_TAB_ID = KEY_EMB_PARENT_TAB_ID;
exports.KEY_EMB_SDK_STARTUP_DURATION = KEY_EMB_SDK_STARTUP_DURATION;
exports.KEY_EMB_SESSION_NUMBER = KEY_EMB_SESSION_NUMBER;
exports.KEY_EMB_SESSION_REASON_ENDED = KEY_EMB_SESSION_REASON_ENDED;
exports.KEY_EMB_SESSION_REASON_STARTED = KEY_EMB_SESSION_REASON_STARTED;
exports.KEY_EMB_STATE = KEY_EMB_STATE;
exports.KEY_EMB_TAB_ID = KEY_EMB_TAB_ID;
exports.KEY_EMB_TYPE = KEY_EMB_TYPE;
exports.KEY_EMB_UNHANDLED_EXCEPTIONS_COUNT = KEY_EMB_UNHANDLED_EXCEPTIONS_COUNT;
exports.KEY_EMB_W3C_TRACEPARENT = KEY_EMB_W3C_TRACEPARENT;
exports.KEY_PREFIX_EMB_PROPERTIES = KEY_PREFIX_EMB_PROPERTIES;
exports.KEY_VIEW_NAME = KEY_VIEW_NAME;
