// @embrace-io/web-sdk/build/cjs/instrumentations/document-load/DocumentLoadInstrumentation/DocumentLoadInstrumentation.cjs
'use strict';

var api = require('@opentelemetry/api');
var core = require('@opentelemetry/core');
var instrumentation = require('@opentelemetry/instrumentation');
var sdkTraceWeb = require('@opentelemetry/sdk-trace-web');
var AttributeNames = require('./enums/AttributeNames.cjs');
var incubating = require('@opentelemetry/semantic-conventions/incubating');
var utils = require('./utils.cjs');
var semanticConventions = require('@opentelemetry/semantic-conventions');
var EmbraceInstrumentationBase = require('../../EmbraceInstrumentationBase/EmbraceInstrumentationBase.cjs');
var attributes = require('../../../constants/attributes.cjs');

const ATTR_DELIVERY_TYPE = 'delivery_type';
const ATTR_ENTRY_TYPE = 'entry_type';
const ATTR_INITIATOR_TYPE = 'initiator_type';
const ATTR_RENDER_BLOCKING_STATUS = 'render_blocking_status';
const ATTR_DECODED_BODY_SIZE = 'decoded_body_size';
class DocumentLoadInstrumentation extends EmbraceInstrumentationBase.EmbraceInstrumentationBase {
    constructor({ diag, perf, enabled, applyCustomAttributesOnSpan, ignorePerformancePaintEvents = false, ignoreNetworkEvents = false } = {}){
        super({
            instrumentationName: 'DocumentLoadInstrumentation',
            instrumentationVersion: '1.0.0',
            diag,
            perf,
            config: {
                enabled,
                applyCustomAttributesOnSpan,
                ignorePerformancePaintEvents,
                ignoreNetworkEvents
            }
        });
        this._onDocumentLoaded = ()=>{
            // Timeout is needed as load event doesn't have yet the performance metrics for loadEnd
            window.setTimeout(()=>{
                this._collectPerformance();
            }, 0);
        };
        if (this._config.enabled) {
            this.enable();
        }
    }
    init() {
        this._diag.debug('Initializing document load instrumentation');
        return undefined;
    }
    /**
   * Adds spans for all resources
   * @param rootSpan
   */ _addResourcesSpans(rootSpan) {
        const resources = performance.getEntriesByType('resource');
        resources.forEach((resource)=>{
            this._initResourceSpan(resource, rootSpan);
        });
    }
    /**
   * Collects information about performance and creates appropriate spans
   */ _collectPerformance() {
        const metaElement = Array.from(document.getElementsByTagName('meta')).find((e)=>e.getAttribute('name') === core.TRACE_PARENT_HEADER);
        const entries = utils.getPerformanceNavigationEntries();
        const traceparent = metaElement && metaElement.content || '';
        api.context.with(api.propagation.extract(api.ROOT_CONTEXT, {
            traceparent
        }), ()=>{
            const rootSpan = this._startSpan(AttributeNames.AttributeNames.DOCUMENT_LOAD, sdkTraceWeb.PerformanceTimingNames.FETCH_START, entries);
            if (!rootSpan) {
                return;
            }
            api.context.with(api.trace.setSpan(api.context.active(), rootSpan), ()=>{
                const fetchSpan = this._startSpan(AttributeNames.AttributeNames.DOCUMENT_FETCH, sdkTraceWeb.PerformanceTimingNames.FETCH_START, entries);
                if (fetchSpan) {
                    fetchSpan.setAttribute(incubating.ATTR_URL_FULL, location.href);
                    api.context.with(api.trace.setSpan(api.context.active(), fetchSpan), ()=>{
                        sdkTraceWeb.addSpanNetworkEvents(fetchSpan, entries, this.getConfig().ignoreNetworkEvents);
                        this._addCustomAttributesOnSpan(fetchSpan, this.getConfig().applyCustomAttributesOnSpan?.documentFetch);
                        this._endSpan(fetchSpan, sdkTraceWeb.PerformanceTimingNames.RESPONSE_END, entries);
                    });
                }
            });
            rootSpan.setAttribute(attributes.KEY_EMB_TYPE, attributes.EMB_TYPES.DocumentLoad);
            rootSpan.setAttribute(incubating.ATTR_URL_FULL, location.href);
            rootSpan.setAttribute(incubating.ATTR_USER_AGENT_ORIGINAL, navigator.userAgent);
            this._addResourcesSpans(rootSpan);
            if (!this.getConfig().ignoreNetworkEvents) {
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.FETCH_START, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.UNLOAD_EVENT_START, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.UNLOAD_EVENT_END, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.DOM_INTERACTIVE, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.DOM_CONTENT_LOADED_EVENT_START, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.DOM_CONTENT_LOADED_EVENT_END, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.DOM_COMPLETE, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.LOAD_EVENT_START, entries);
                sdkTraceWeb.addSpanNetworkEvent(rootSpan, sdkTraceWeb.PerformanceTimingNames.LOAD_EVENT_END, entries);
            }
            if (!this.getConfig().ignorePerformancePaintEvents) {
                utils.addSpanPerformancePaintEvents(rootSpan);
            }
            this._addCustomAttributesOnSpan(rootSpan, this.getConfig().applyCustomAttributesOnSpan?.documentLoad);
            this._endSpan(rootSpan, sdkTraceWeb.PerformanceTimingNames.LOAD_EVENT_END, entries);
        });
    }
    /**
   * Helper function for ending span
   * @param span
   * @param performanceName name of performance entry for time end
   * @param entries
   */ // eslint-disable-next-line @typescript-eslint/class-methods-use-this
    _endSpan(span, performanceName, entries) {
        // span can be undefined when entries are missing the certain performance - the span will not be created
        if (span) {
            if (sdkTraceWeb.hasKey(entries, performanceName) && typeof entries[performanceName] === 'number') {
                span.end(entries[performanceName]);
            } else {
                span.end();
            }
        }
    }
    /**
   * Creates and ends a span with network information about resource added as timed events
   * @param resource
   * @param parentSpan
   */ _initResourceSpan(resource, parentSpan) {
        const span = this._startSpan(AttributeNames.AttributeNames.RESOURCE_FETCH, sdkTraceWeb.PerformanceTimingNames.FETCH_START, resource, parentSpan);
        if (span) {
            span.setAttribute(attributes.KEY_EMB_TYPE, attributes.EMB_TYPES.ResourceFetch);
            span.setAttribute(incubating.ATTR_URL_FULL, resource.name);
            sdkTraceWeb.addSpanNetworkEvents(span, resource, this.getConfig().ignoreNetworkEvents);
            if (resource.deliveryType) {
                span.setAttribute(ATTR_DELIVERY_TYPE, resource.deliveryType);
            }
            span.setAttribute(ATTR_ENTRY_TYPE, resource.entryType);
            span.setAttribute(ATTR_INITIATOR_TYPE, resource.initiatorType);
            if (resource.renderBlockingStatus) {
                span.setAttribute(ATTR_RENDER_BLOCKING_STATUS, resource.renderBlockingStatus);
            }
            span.setAttribute(semanticConventions.ATTR_HTTP_RESPONSE_STATUS_CODE, resource.responseStatus);
            span.setAttribute(incubating.ATTR_HTTP_RESPONSE_BODY_SIZE, resource.encodedBodySize);
            span.setAttribute(incubating.ATTR_HTTP_RESPONSE_SIZE, resource.transferSize);
            span.setAttribute(ATTR_DECODED_BODY_SIZE, resource.decodedBodySize);
            this._addCustomAttributesOnResourceSpan(span, resource, this.getConfig().applyCustomAttributesOnSpan?.resourceFetch);
            this._endSpan(span, sdkTraceWeb.PerformanceTimingNames.RESPONSE_END, resource);
        }
    }
    /**
   * Helper function for starting a span
   * @param spanName name of span
   * @param performanceName name of performance entry for time start
   * @param entries
   * @param parentSpan
   */ _startSpan(spanName, performanceName, entries, parentSpan) {
        if (sdkTraceWeb.hasKey(entries, performanceName) && typeof entries[performanceName] === 'number') {
            const span = this.tracer.startSpan(spanName, {
                startTime: entries[performanceName]
            }, parentSpan ? api.trace.setSpan(api.context.active(), parentSpan) : undefined);
            return span;
        }
        return undefined;
    }
    /**
   * executes callback {_onDocumentLoaded} when the page is loaded
   */ _waitForPageLoad() {
        if (window.document.readyState === 'complete') {
            this._onDocumentLoaded();
        } else {
            window.addEventListener('load', this._onDocumentLoaded);
        }
    }
    /**
   * adds custom attributes to root span if configured
   */ _addCustomAttributesOnSpan(span, applyCustomAttributesOnSpan) {
        if (applyCustomAttributesOnSpan) {
            instrumentation.safeExecuteInTheMiddle(()=>{
                applyCustomAttributesOnSpan(span);
            }, (error)=>{
                if (!error) {
                    return;
                }
                this._diag.error('addCustomAttributesOnSpan', error);
            }, true);
        }
    }
    /**
   * adds custom attributes to span if configured
   */ _addCustomAttributesOnResourceSpan(span, resource, applyCustomAttributesOnSpan) {
        if (applyCustomAttributesOnSpan) {
            instrumentation.safeExecuteInTheMiddle(()=>{
                applyCustomAttributesOnSpan(span, resource);
            }, (error)=>{
                if (!error) {
                    return;
                }
                this._diag.error('addCustomAttributesOnResourceSpan', error);
            }, true);
        }
    }
    enable() {
        window.removeEventListener('load', this._onDocumentLoaded);
        this._waitForPageLoad();
    }
    disable() {
        window.removeEventListener('load', this._onDocumentLoaded);
    }
}

exports.DocumentLoadInstrumentation = DocumentLoadInstrumentation;
