// @embrace-io/web-sdk/build/esm/api-sessions/api/SessionAPI/SessionAPI.js
import { ProxySpanSessionManager } from '../../manager/ProxySpanSessionManager/ProxySpanSessionManager.js';

class SessionAPI {
    constructor({ proxySpanSessionManager }){
        this.getSpanSessionManager = ()=>{
            return this._proxySpanSessionManager;
        };
        this._proxySpanSessionManager = proxySpanSessionManager;
    }
    static getInstance() {
        if (!SessionAPI._instance) {
            SessionAPI._instance = new SessionAPI({
                proxySpanSessionManager: new ProxySpanSessionManager()
            });
        }
        return SessionAPI._instance;
    }
    setGlobalSessionManager(sessionManager) {
        this._proxySpanSessionManager.setDelegate(sessionManager);
    }
    addBreadcrumb(name) {
        this.getSpanSessionManager().addBreadcrumb(name);
    }
    addProperty(key, value, options) {
        this.getSpanSessionManager().addProperty(key, value, options);
    }
    removeProperty(key) {
        this.getSpanSessionManager().removeProperty(key);
    }
    endSessionSpan() {
        this.getSpanSessionManager().endSessionSpan();
    }
    endSessionSpanInternal(reason) {
        this.getSpanSessionManager().endSessionSpanInternal(reason);
    }
    currentSessionAsReadableSpan(reason) {
        return this.getSpanSessionManager().currentSessionAsReadableSpan(reason);
    }
    getSessionId() {
        return this.getSpanSessionManager().getSessionId();
    }
    getSessionSpan() {
        return this.getSpanSessionManager().getSessionSpan();
    }
    getSessionStartTime() {
        return this.getSpanSessionManager().getSessionStartTime();
    }
    startSessionSpan(options) {
        this.getSpanSessionManager().startSessionSpan(options);
    }
    addSessionStartedListener(listener) {
        return this.getSpanSessionManager().addSessionStartedListener(listener);
    }
    addSessionEndedListener(listener) {
        return this.getSpanSessionManager().addSessionEndedListener(listener);
    }
}

export { SessionAPI };
