// @embrace-io/web-sdk/build/cjs/instrumentations/exceptions/react/EmbraceErrorBoundary/EmbraceErrorBoundary.cjs
'use strict';

var React = require('react');
var logAPI = require('../../../../api-logs/logAPI.cjs');
var attributes = require('../../../../constants/attributes.cjs');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

class EmbraceErrorBoundary extends React__namespace.Component {
    constructor(props){
        super(props);
        this._logManager = logAPI.log.getLogManager();
        this.state = {
            hasError: false
        };
    }
    static getDerivedStateFromError() {
        return {
            hasError: true
        };
    }
    componentDidCatch(error, errorInfo) {
        this._logManager.logException(error, {
            handled: false,
            attributes: {
                'react.component_stack': errorInfo.componentStack ?? undefined,
                [attributes.KEY_EMB_INSTRUMENTATION]: attributes.EMB_ERROR_INSTRUMENTATIONS.ReactErrorBoundary
            }
        });
    }
    render() {
        if (this.state.hasError) {
            return this.props.fallback();
        }
        return this.props.children;
    }
}

exports.EmbraceErrorBoundary = EmbraceErrorBoundary;
